File: data/python/src/function_numpy_divide.py
import numpy as np


class DivisionCalculator:

    def divide(self, arr1, arr2=None, scalar=None):
        try:
            if arr2 is not None:  # 两个数组相除
                # 确保输入为基本数据类型（int、float）
                if not isinstance(arr1, (int, float)) or not isinstance(arr2, (int, float)):
                    raise ValueError("Both arr1 and arr2 must be basic data types (int or float).")

                # 进行除法运算
                result = arr1 / arr2
                return result  # 返回结果为基本数据类型（float）

            elif scalar is not None:  # 数组与标量相除
                if not isinstance(arr1, (int, float)):
                    raise ValueError("arr1 must be a basic data type (int or float) when dividing by scalar.")

                # 进行除法运算
                result = arr1 / scalar
                return result  # 返回结果为基本数据类型（float）

            else:
                raise ValueError("Either arr2 or scalar must be provided.")

        except Exception as e:
            raise ValueError(f"Error in division: {e}")

File: data/cpp/src/function_numpy_divide.cpp
#include <string>
#include <sstream>
#include <optional>
#include <stdexcept>
#include <xtensor/xarray.hpp>
#include <xtensor/xio.hpp>

class DivisionCalculator
{
public:
    template <typename T>
    double divide(const T& arr1,
                  const std::optional<T>& arr2 = std::nullopt,
                  const std::optional<T>& scalar = std::nullopt)
    {
        try
        {
            // 如果 arr2 有值，则执行 arr1 / arr2
            if (arr2.has_value())
            {
                xt::xarray<T> A = {arr1};
                xt::xarray<T> B = {arr2.value()};
                auto result = A / B;  // 计算除法
                return static_cast<double>(result(0));
            }
            // 如果提供了 scalar，则执行 arr1 / scalar
            else if (scalar.has_value())
            {
                xt::xarray<T> A = {arr1};
                xt::xarray<T> S = {scalar.value()};
                auto result = A / S;
                return static_cast<double>(result(0));
            }
            // 否则抛出错误
            else
            {
                throw std::runtime_error("Either arr2 or scalar must be provided.");
            }
        }
        catch(const std::exception& e)
        {
            throw std::runtime_error(std::string("Error in division: ") + e.what());
        }
    }
};